### ISSUE TITLE
Limit rules to certain time during the day

### ISSUE THREAD

**Issue by jurgenhaas** [Yelp/elastalert#492]
We have a few cardinality rules defined that are checking that our services are running OK by making sure that a certain number of events can be found in ES. Now, some of those rules only make sense within business hours and we would love to define such periods as part of the rules.

In other words, let's enable some rules to be executed only within specific periods each day.

Is that a valid feature request?

**Comment by dmaciel** [comment]
I just trying to do this using a new rule type, but I will be better if this use case is supported by the app by default

**Comment by lwhitworth** [comment]
Just came here to request the same.  In my scenario I want to use a flatline rule to alert if succesful auths to our Radius server stop working (so no successful auths in x minutes).  However once everyone goes home for the evening we see a few hours of no auths so the rule would be triggered throughout the night and at weekends.  What I'd like is in the rule the ability to say only alert from 9am to 5pm.

Only solution I can see at the moment is to alert out to a third party that handles this part of the logic, but native support would be awesome

**Comment by jurgenhaas** [comment]
Just thought of a workaround here: we could move the relevant rule files out of the rules directory at 5pm with a cron task and move them back into that directory at 9am. As far as I understand, ElastAlert can already handle dynamic changes in the rules directory. I probably give that a try until we come to a conclusion on how to proceed with this feature request.

**Comment by Qmando** [comment]
If you do that, be sure to set ``old_query_limit`` to something small in config.yaml, otherwise, when you add the rule back, it will pick up from where it left off the previous day. It will always pick up where it last left off for any rule, given that it's within ``old_query_limit``, which defaults to 1 week.

**Comment by BaeHwidong-NBT** [comment]
I reached here thinking of the exactly same feature as @lwhitworth wrote. I'll try @jurgenhaas's suggestion for now. Thank you!

**Comment by iekulyk** [comment]
You can configure you filter to ignore records within certain timespans

**Comment by JustinPealing** [comment]
Could the `run_every` option is enhanced to accept cron syntax in the same way that the 'aggregation' option does? e.g:

```
run_every:
    schedule: '* 9-17 * * 1-5'
```

**Comment by Qmando** [comment]
Yes. I'm sorry this has been outstanding for so long. I will try to finish it up soon.

**Comment by josephka333** [comment]
Hi @Qmando please do help us with this feature its very useful and would be a great addition.

**Comment by pvsms** [comment]
I landed here for the very same request to run the rule only during certain hours of the day. Let me know if there is any feature enhanced on the same.

**Comment by joeyJsonar** [comment]
@Qmando Hello, we managed to integrate python [APScheduler](https://apscheduler.readthedocs.io/en/latest/) on our fork of elastalert. I just didn't have the time to have a pull request. I'll schedule to do one this Friday or weekend (long weekend here in Canada so I should be able to find the time).

**Comment by pvsms** [comment]
Excellent! How quickly the PR can be merged and get a new JAR version, just to understand when i can integrate with it and plan my activities accordingly. Thanks again!

**Comment by husinul** [comment]
we tried with new rule type , it works fine, but default would be great.

**Comment by pvsms** [comment]
@joeyJsonar: Were able to create and get the PR merged? Thx.

**Comment by Qmando** [comment]
I've merged this feature into a new branch, `beta`, and released it as a new package version `0.2.0b1` available on pypi. 

This includes a couple other changes as well, like threading support, but you can now limit rule execution to certain times of the day using `limit_execution` using cron syntax. For example
```
limit_execution: "* 7-22 * * *"
```
Would mean to only run the rule between 7 am and 10 pm every day.

This feature is still in beta, of course, but you're welcome to try.

**Comment by JaredRietdyk** [comment]
@Qmando, I've been running with limit_execution for a few days now and it seems to be working well. I believe there might be an issue with the frequency type, where after the cron is done it will go back and alert on the results since the last cron.

**Comment by gsagwan** [comment]
Hi,
i am facing issue where in my rule is paused until the cron specified.
limit_execution: "* 17-23 * * *"
Any idea ?

**Comment by JaredRietdyk** [comment]
@gsagwan, I was confused at first as well. The limit execution is limiting the rule to run ONLY during that time. So if you want the rule to run and stop during that time you would need to use something like:
limit_execution:  "* 0-16,24 * * *"

**Comment by gsagwan** [comment]
> @gsagwan, I was confused at first as well. The limit execution is limiting the rule to run ONLY during that time. So if you want the rule to run and stop during that time you would need to use something like:
> limit_execution: "* 0-16,24 * * *"

Hi @JaredRietdyk , Thanks for your comments.
I am trying to put the below cron.
limit_execution: "* 04-17 * * *"

But it continues to run post 17 as well.
Not able to figure out how this is working.

**Comment by thesm3rdo** [comment]
Hi @gsagwan , is there an update on this limit_execution property ? I cannot see it in the beta branch and am wanting to track its progress as it's a piece of functionality that will be very useful to our business.

**Comment by Qmando** [comment]
@thesm3rdo 
https://github.com/Yelp/elastalert/blob/beta/elastalert/elastalert.py#L1238

It's there for use already. No update regarding using local timezones, or when this will be released from beta to a full release.

**Comment by Atem18** [comment]
@Qmando I see that you are pushing things to Master branch but not Beta.
So what is the current status of the feature ?
Do you cherry-pick commits from beta branch to master ?

**Comment by Qmando** [comment]
I created the beta release for testing some large and potentially dangerous changes, in order to get some help testing them. When I merge it into a new release, it will contain everything from the master branch too. Apologies for the very slow pace of releases.

**Comment by Atem18** [comment]
@Qmando No problem, make the best production release possible ! :)

**Comment by Atem18** [comment]
@Qmando Hi, any update on merging in to master ?

**Comment by Qmando** [comment]
I did pull the changes form master into beta. But, I'll probably wait until I've finished the py3 migration to release those changes. Sorry It's been very slow.

**Comment by Atem18** [comment]
Ok thanks a lot

**Comment by 0x-29A** [comment]
I found a  solution making an Enhancement so that you can "plug it" on any rule of yours : 
What do you think about it ? @Qmando 

https://github.com/0x-29A/elastalert_hour_range

**Comment by anuarabdullah** [comment]
Hi, as today can I know how to implement the limit_execution function in elastalert properly? I use limit_execution: "* 6-23 * * *" in my rule.yml but the rule keep running past 2300 hrs.

Also this limit_execution run on local timezone or other timezone?

**Comment by rabiashaikabdulkader** [comment]
> 0.2.0b1

Hi @Qmando I would like to ask if this feature is available in elastalert 0.2.2 version. I am looking to check for the rules at a particular time and trigger an alert. May I know how I should achieve it? Can you please guide me by copy pasting the sample alert.yaml file

**Comment by rabiashaikabdulkader** [comment]
> I've merged this feature into a new branch, `beta`, and released it as a new package version `0.2.0b1` available on pypi.
> 
> This includes a couple other changes as well, like threading support, but you can now limit rule execution to certain times of the day using `limit_execution` using cron syntax. For example
> 
> ```
> limit_execution: "* 7-22 * * *"
> ```
> 
> Would mean to only run the rule between 7 am and 10 pm every day.
> 
> This feature is still in beta, of course, but you're welcome to try.

Hi @Qmando I would like to ask if this feature is available in elastalert 0.2.2 version. I am looking to check for the rules at a particular time and trigger an alert. May I know how I should achieve it? Can you please guide me by copy pasting the sample alert.yaml file

**Comment by svenkyedem** [comment]
HI @Qmando . I have tested limit_execution feature.i have given cron as below
limit_execution: "55-55 6-7 * * *"
In logs it is showing like below
INFO:elastalert:Pausing Backup Fail Alert rule until next run at 2020-11-11 06:55 UTC
When the time is 6:55 UTC it is showing like
INFO:elastalert:Pausing Backup Fail Alert rule until next run at 2020-11-11 07:55 UTC
When the time is 7:55 UTC it is showing like 
INFO:elastalert:Pausing Backup Fail Alert rule until next run at 2020-11-12 06:55 UTC

Basically it is not executing anytime.

elastalert version: v0.2.0

**Comment by jgutta** [comment]
Hi! Also wanted to express interest in this feature. It doesn't seem to be in V0.2.4, is there any a plan to enable it soon?

**Comment by r1296** [comment]
Would be great if this feature could be added in one of the next releases.

**Comment by nsano-rururu** [comment]
@r1296 

As far as the following issues are seen, it seems that the feature addition is not released quickly. Let's keep an eye on the future situation.

[Please don't let this die. ElastAlert usecase is very strong.](https://github.com/Yelp/elastalert/issues/2947)
[Should someone else maintain this?](https://github.com/Yelp/elastalert/issues/2946)
[Status of the elastalert project](https://github.com/Yelp/elastalert/issues/2911)

**Comment by ParshantMehra** [comment]
Any update when will this feature be available?

**Comment by fberrez** [comment]
elastalert is no more maintained. 
However, the `limit_execution` feature was added in elastalert 0.2.0 (see: https://github.com/jertel/elastalert2/discussions/106) but the documentation has not been updated.
